benchmarks/halo_finder/ArborX_HaloFinder.hpp

```diff
@@ -0,0 +1,264 @@
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_HALO_FINDER_HPP
+#define ARBORX_HALO_FINDER_HPP
+
+#include <ArborX_DetailsSortUtils.hpp>
+#include <ArborX_DetailsUtils.hpp>
+#include <ArborX_LinearBVH.hpp>
+#include <ArborX_Macros.hpp>
+
+#include <chrono>
+
+namespace ArborX
+{
+
+template <typename View>
+struct Wrapped
+{
+  View _M_view;
+  double _r;
+};
+
+template <typename View>
+auto wrap(View v, double r)
+{
+  return Wrapped<View>{v, r};
+}
+
+namespace Traits
+{
+template <typename View>
+struct Access<Wrapped<View>, PredicatesTag>
+{
+  using memory_space = typename View::memory_space;
+  static size_t size(Wrapped<View> const &w) { return w._M_view.extent(0); }
+  static KOKKOS_FUNCTION auto get(Wrapped<View> const &w, size_t i)
+  {
+    return attach(intersects(Sphere{w._M_view(i), w._r}), (int)i);
+  }
+};
+} // namespace Traits
+
+namespace HaloFinder
+{
+template <typename MemorySpace>
+struct CCSCallback
+{
+  using tag = ArborX::Details::InlineCallbackTag;
+  Kokkos::View<int *, MemorySpace> stat_;
+
+  KOKKOS_INLINE_FUNCTION
+  int representative(int const i) const
+  {
+    // ##### ECL license (see LICENSE.ECL) #####
+    int curr = stat_(i);
+    if (curr != i)
+    {
+      int next, prev = i;
+      while (curr > (next = stat_(curr)))
+      {
+        stat_(prev) = next;
+        prev = curr;
+        curr = next;
+      }
+    }
+    return curr;
+  }
+
+  template <typename Query, typename Insert>
+  KOKKOS_FUNCTION void operator()(Query const &query, int j,
+                                  Insert const &) const
+  {
+    int const i = ArborX::getData(query);
+
+    // only process edge in one direction
+    if (i > j)
+    {
+      // initialize to the first neighbor that's smaller
+      if (Kokkos::atomic_compare_exchange(&stat_(i), i, j) == i)
+        return;
+
+      {
+        // ##### ECL license (see LICENSE.ECL) #####
+        int vstat = representative(i);
+        int ostat = representative(j);
+
+        bool repeat;
+        do
+        {
+          repeat = false;
+          if (vstat != ostat)
+          {
+            int ret;
+            if (vstat < ostat)
+            {
+              if ((ret = Kokkos::atomic_compare_exchange(&stat_(ostat), ostat,
+                                                         vstat)) != ostat)
+              {
+                ostat = ret;
+                repeat = true;
+              }
+            }
+            else
+            {
+              if ((ret = Kokkos::atomic_compare_exchange(&stat_(vstat), vstat,
+                                                         ostat)) != vstat)
+              {
+                vstat = ret;
+                repeat = true;
+              }
+            }
+          }
+        } while (repeat);
```

Comment:
```suggestion
if(vstat < ostat)
  while(Kokkos::atomic_compare_exchange_strong(&stat_(ostat), ostat, vstat)
    ;
else
  while(Kokkos::atomic_compare_exchange_strong(&stat_(vstat), vstat, ostat)
    ;

Reply:
No, I don't think these are equivalent. And in general, this code is very sensitive and taken whole from ECL-CC. I don't want to touch it.

Reply:
No, they are not equivalent.

Reply:
Yep I see it now

Reply:
You can still use `atomic_compare_exchange_strong()` though

Reply:
Sure, but I don't want to modify the code without a strong need, and prefer to leave it in the original ECL-CC form.

Reply:
Actually how about 
```C++
if(vstat < ostat)
  while(Kokkos::atomic_compare_exchange_strong(&stat_(ostat), ostat, vstat)
    ostat = vstat;
else
  while(Kokkos::atomic_compare_exchange_strong(&stat_(vstat), vstat, ostat)
    vstat = ostat;
```

Reply:
This is still different. Say you are in the `if()` branch. If `vstat` is updated by a different thread, you won't notice it in your code, but will in the original.